modopt/opt/algorithms/gradient_descent.py

```diff
@@ -0,0 +1,354 @@
+        else:
+            self._cost_func = cost
+        # Set the algorithm parameters
+        for param_val in (eta, epsilon):
+            self._check_param(param_val)
+        self._eta = eta
+        self._eps = epsilon
+
+        # Set the algorithm parameter update methods
+        self._check_param_update(eta_update)
+        self._eta_update = eta_update
+        self.idx = 0
+        self.epoch_size = epoch_size
+
+
+    def iterate(self, max_iter=150):
+        """Iterate.
+
+        This method calls update until either convergence criteria is met or
+        the maximum number of iterations is reached.
+
+        Parameters
+        ----------
+        max_iter : int, optional
+            Maximum number of iterations (default is ``150``)
+
+        """
+        self._run_alg(max_iter)
+
+        # retrieve metrics results
+        self.retrieve_outputs()
+
+        self.x_final = self._x_new
+
+    def _update(self):
+        self._grad.get_grad(self._x_old)
+        self.update_grad_dir(self._grad.grad)
+        self.update_grad_speed(self._grad.grad)
+        step = self._eta / (np.sqrt(self._speed_grad) + self._eps)
+        self._x_new = self._x_old - step * self._dir_grad
+        if self.idx % self.epoch_size == 0:
+            self.reset()
+            self.update_reg(step)
+        self._x_old = self._x_new.copy()
+        if self._eta_update is not None:
+            self._eta = self._eta_update(self._eta, self.idx)
+        # Test cost function for convergence.
+        if self._cost_func:
+            self.converge = (
+                self.any_convergence_flag()
+                or self._cost_func.get_cost(self._x_new)
+            )
+
+    def update_grad_dir(self, grad):
+        """Update the gradient descent direction."""
+        self._dir_grad = grad
+
+    def update_grad_speed(self, grad):
+        """Update the gradient descent speed."""
+        pass
+
+    def update_reg(self, factor):
+        """Regularisation step."""
+        self._x_new = self._prox.op(self._x_new, extra_factor=factor)
+
+    def get_notify_observers_kwargs(self):
+        """Notify observers.
+
+        Return the mapping between the metrics call and the iterated
+        variables.
+
+        Returns
+        -------
+        notify_observers_kwargs : dict,
+           The mapping between the iterated variables
+
+        """
+        return {
+            'x_new': self._x_new,
+            'dir_grad': self._dir_grad,
+            'speed_grad': self._speed_grad,
+            'idx': self.idx,
+        }
+
+    def retrieve_outputs(self):
+        """Retrieve outputs.
+
+        Declare the outputs of the algorithms as attributes: x_final,
+        y_final, metrics.
+
+        """
+        metrics = {}
+        for obs in self._observers['cv_metrics']:
+            metrics[obs.name] = obs.retrieve_metrics()
+        self.metrics = metrics
+
+    def reset(self):
+        """Reset internal state of the algorithm."""
+        pass
```

Comment:
Same point regarding method overriding.

Reply:
same answer ;) some class reimplements it (AdaGrad for example), other not.